### PR TITLE
[codegen/nodejs] - Fix provider enum with env var

### LINKED ISSUE #8042: [nodejs] Provider enum with env vars generates invalid code
## Steps to reproduce

The following schema is generating invalid TypeScript code:

```
"region": {
    "oneOf": [
        {
            "type": "string"
        },
        {
            "$ref": "#/types/aws-native:index/region:Region"
        }
    ],
    "description": "The region where AWS operations will take place. Examples are `us-east-1`, `us-west-2`, etc."
},
```

Expected: 
I'm not 100% sure, but I think it should look like this instead.
```ts
inputs["region"] = (args ? args.region : undefined) ?? utilities.getEnv("AWS_REGION", "AWS_DEFAULT_REGION");

```
Actual: 
```ts
inputs["region"] = pulumi.output((args ? args.region : undefined) ?? <any>utilities.getEnv("AWS_REGION", "AWS_DEFAULT_REGION")).apply(JSON.stringify);

```

Running an update with this code results in the following error:
```
error: Configuration 'aws-native:region' value 'us-west-2' is not a valid JSON object
```

### PR DESCRIPTION
**Pull request by komalali**
# Description

Fixes #8042 

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Review by lblackstone — APPROVED** [review]
Wow, that was fast! Thanks!

File: pkg/codegen/nodejs/gen.go [hunk 1]
```diff
@@ -327,8 +327,18 @@ func (mod *modContext) typeString(t schema.Type, input bool, constValue interfac
 func isStringType(t schema.Type) bool {
 	t = codegen.UnwrapType(t)
 
-	for tt, ok := t.(*schema.TokenType); ok; tt, ok = t.(*schema.TokenType) {
-		t = tt.UnderlyingType
+	switch typ := t.(type) {
+	case *schema.TokenType:
+		t = typ.UnderlyingType
+	case *schema.UnionType:
+		for _, tt := range typ.ElementTypes {
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen.go:334]: This is cryptic. Loop through elements of the Union Type, filter them to Enum Type, get the *last* ElementType of the last element that is Enum Type, and compare to string.

Comment (pgavlin) [reply pkg/codegen/nodejs/gen.go:334]: Wondering about this myself--I think that I recognize the pattern, but I've forgotten exactly what the pattern is. @komalali could you elaborate on this bit?

**Review by pgavlin — COMMENTED** [review]
LGTM--just one question re: clarifying a case

File: pkg/codegen/nodejs/gen.go (hunk 1 above)
Comment (komalali) [reply pkg/codegen/nodejs/gen.go:334]: This case is for relaxed enums that are represented as the union of the primitive type and the enum type.
```json
"oneOf": [
  {
    "type": "string"
  },
  {
    "$ref": "#/types/configstation:index:color"
  }
]
```

 The enum type itself is:
```json
"configstation:index:color": {
  "enum": [
    {"value": "blue"},
    {"value": "red"}
  ],
  "type": "string"
}
```

I chose this pattern as if it's a Union, check if one of the elements is an Enum (this signifies a relaxed enum) and then `t` is the `ElementType` of the `Enum`.

Comment (komalali) [reply pkg/codegen/nodejs/gen.go:334]: I can add a comment to that effect if it's helpful to clarify.

Comment (t0yv0) [reply pkg/codegen/nodejs/gen.go:334]: Thank you that'd be much appreciated by the future readers, a comment, or a comment + unit test if so inspired. Thanks.

**Review by t0yv0 — APPROVED** [review]

File: pkg/codegen/nodejs/gen.go (hunk 1 above)
Comment (komalali) [reply pkg/codegen/nodejs/gen.go:334]: Where would you like me to add the unit test? My test case already included in this PR addresses this specific scenario.

File: pkg/codegen/internal/test/testdata/provider-config-schema/schema.json [hunk 2]
```diff
@@ -1,6 +1,23 @@
 {
   "version": "0.0.1",
   "name": "configstation",
+  "provider": {
+    "inputProperties": {
+      "favoriteColor": {
```
Comment (pgavlin) [inline pkg/codegen/internal/test/testdata/provider-config-schema/schema.json:6]: Maybe add a description here explaining that this is a relaxed enum?

File: pkg/codegen/nodejs/gen.go (hunk 1 above)
Comment (pgavlin) [reply pkg/codegen/nodejs/gen.go:334]: > I can add a comment to that effect if it's helpful to clarify.

I think a comment here and perhaps in the description of the relaxed enum field would be sufficient 👍

**Review by pgavlin — APPROVED** [review]

File: pkg/codegen/nodejs/gen.go (hunk 1 above)
Comment (t0yv0) [reply pkg/codegen/nodejs/gen.go:334]: Not asking for coverage but clarification. We don't want the reader to ponder what this does. If you're adding the comment  that's fine too. As a reader if I see isString() in gen.go I don't understand I look for gen_test.go to see maybe there's tests covering isString with clarifying examples and comments that tell the story what this does.

Comment (komalali) [reply pkg/codegen/nodejs/gen.go:338]: You got all the things

**Comment by komalali** [comment]
Unrelated test failure, merging

**Comment by t0yv0** [comment]
Collected into flaky list: https://github.com/pulumi/pulumi/issues/8061

**Pull request merged**